pilot/pkg/networking/plugin/httpmx/httpmx.go

```diff
@@ -0,0 +1,109 @@
+type Plugin struct{}
+
+// NewPlugin returns an instance of the httpmx plugin
+func NewPlugin() plugin.Plugin {
+	return Plugin{}
+}
+
+// OnInboundListener is called whenever a new listener is added to the LDS output for a given service
+// Can be used to add additional filters.
+func (p Plugin) OnInboundListener(in *plugin.InputParams, mutable *networking.MutableObjects) error {
+	if in.Node.Type != model.SidecarProxy {
+		// Only care about sidecar.
+		return nil
+	}
+	return buildFilter(in, mutable, false)
+}
+
+// OnInboundFilterChains is called whenever a plugin needs to setup the filter chains, including relevant filter chain
+// configuration, like FilterChainMatch and TLSContext.
+func (p Plugin) OnInboundFilterChains(in *plugin.InputParams) []networking.FilterChain {
+	httpmxFilter := buildHTTPMxFilter()
+	fc := networking.FilterChain{HTTP: []*http_conn.HttpFilter{httpmxFilter}}
+	return []networking.FilterChain{fc}
+}
+
+// OnOutboundListener is called whenever a new outbound listener is added to the LDS output for a given service.
+// Can be used to add additional filters on the outbound path.
+func (p Plugin) OnOutboundListener(in *plugin.InputParams, mutable *networking.MutableObjects) error {
+	return nil
+}
+
+// OnInboundPassthrough is called whenever a new passthrough filter chain is added to the LDS output.
+// Can be used to add additional filters.
+func (p Plugin) OnInboundPassthrough(in *plugin.InputParams, mutable *networking.MutableObjects) error {
+	return nil
+}
+
+// OnInboundPassthroughFilterChains is called whenever a plugin needs to setup custom pass through filter chain.
+func (p Plugin) OnInboundPassthroughFilterChains(in *plugin.InputParams) []networking.FilterChain {
+	return nil
+}
+
+func buildFilter(in *plugin.InputParams, mutable *networking.MutableObjects, isPassthrough bool) error {
+	for i := range mutable.FilterChains {
+		// todo: if wasm enabled or not?
+		// switch tpc protocol?
+		if in.ListenerProtocol == networking.ListenerProtocolHTTP || mutable.FilterChains[i].ListenerProtocol == networking.ListenerProtocolHTTP {
+			// if wasm not enabled: build pure http mx
+			if httpMxFilter := buildHTTPMxFilter(); httpMxFilter != nil {
+				mutable.FilterChains[i].HTTP = append(mutable.FilterChains[i].HTTP, httpMxFilter)
+			}
+		}
+	}
+
+	return nil
+}
+
+func buildHTTPMxFilter() *http_conn.HttpFilter {
+	inlineString := "envoy.wasm.metadata_exchange"
+	dataSourceInlineString := &v3.DataSource_InlineString{
+		InlineString: inlineString,
+	}
+
+	code := &v3.AsyncDataSource_Local{
+		Local: &v3.DataSource{
+			Specifier: dataSourceInlineString,
+		}}
+	vmConfig := &wasm_v3.VmConfig{
+		Runtime: "envoy.wasm.runtime.null",
+		Code:    &v3.AsyncDataSource{Specifier: code},
+	}
+
+	pluginisVM := &wasm_v3.PluginConfig_VmConfig{
+		VmConfig: vmConfig,
+	}
+
+	wasmPluginConfig := &wasm_v3.PluginConfig{
+		Vm: pluginisVM,
+		//Configuration: util.MessageToAny(),
+	}
+	filterConfigProto := &wasm.Wasm{
```

Comment:
nit: this may be more clear to make it a big inline config object, as is its hard to understand the full shape

Reply:
This should be all static and perhaps a static variable? not sure if the compiler is smart enough to do it...